File: src/core/utilities/Result.cpp
#include <core/utilities/Result.hpp>

#ifndef HYP_BUILDTOOL
#include <Result.generated.inl>
#endif

namespace Hyperion {
namespace utilities {

class NullError final : public Error
{
public:
    NullError() = default;
};

HYP_API const Error& GetNullError()
{
    static NullError s_nullError;

    return s_nullError;
}

} // namespace utilities
} // namespace Hyperion

File: Source/Engine/DotNET/Shared/Utilities/Result.cs
using System;
using System.Runtime.InteropServices;

namespace Hyperion
{
    // Maps to core/utilities/Result.hpp
    [ClassBinding(Name="Result")]
    [StructLayout(LayoutKind.Explicit, Size = 8)]
    public struct Result
    {
        public static Result OK => new Result();

        [FieldOffset(0)]
        private PimplPtr _error;

        public Result()
        {
        }

        public bool IsValid
        {
            get
            {
                return _error.ptr == IntPtr.Zero;
            }
        }
    }
}